docs/how-packages-and-artifacts-are-built.md

```diff
@@ -0,0 +1,114 @@
+```
+The the "exclude" field is to say that the package `"dcos-installer-ui"` should not be built in the default variant. The `"bootstrap_package_list"` is a whitelist of packages that will be placed in the default bootstrap tarball. The bootstrap tarball is the collection of packages transferred to host machines which then orchestrates the bootstrapping of the remaining packages to spin up DC/OS. As such, the installer UI code is completely useless to a DC/OS deployment.
+
+Now, if we consider [installer.treeinfo.json](../packages/installer.treeinfo.json):
+```
+{
+  "core_package_list": [
+    "dcos-image",
+    "dcos-installer-ui"
+  ]
+}
+```
+The `"core_package_list"` field declares a whitelist of packages that are to be built and packaged into the bootstrap tarball for the variant `installer`.
+
+The purpose of the above configuration is to generate two bootstrap tarballs with each build. The default tarball will contain the code that handles the import of the remaining DC/OS packages (as dictated by the provider `gen.build_deploy` module) and setup the hosts to startup DC/OS. The installer tarball will be used to create a mock DC/OS environment which can provide packages to hosts via the `dcos_installer` program (see below).
+
+#### buildinfo.json
+Variants can also allow for customized packages. In any given package, one may add a `foobar.buildinfo.json` that will pull in alternative sources to run against the same `build` script. To have a package variant built, the `treeinfo.json` must have a `"variant"` section that describes the specific package variant to be used with the tree variant. This logic must be done explicitly as there is no implicit matching between package and tree variants.
+
+#### upstream.json
+The other file that can be placed in the `packages` directory is called `upstream.json`. This JSON is formatted like a single source from the aforementioned `buildinfo.json`. Here's an example:
+```
+{
+  "git": "https://github.com/dcos/dcos",
+  "kind": "git",
+  "ref": "9364e0a72b942e8dcec306ef7c1673c6f23bd9cc",
+  "ref_origin": "master"
+}
+```
+This file will tell pkgpanda the packages directory containing it should be expanded to include the packages from the given upstream source. Thus, one may create a modified version of DC/OS by just defining the desired package modifications; no forked repositories required!
+
+### Developing Packages
+When building a release, `pkgpanda` is only called through its library. However, there is a CLI which is very helpful for developing and debugging packages. To setup pkgpanda locally, be sure docker is installed and able to run without sudo, then run the `./prep_local` script in the root directory (python virtualenv will be required). Finally, make sure the dcos-builder image has been prepped locally by doing `cd pkgpanda/build/docker/dcos-builder/; docker build -t dcos/dcos-builder:dcos-builder_dockerdir-latest .` Right now, this is handled apart of the release scripting, not pkgpanda.
```

Comment:
For the compound command can we do `&&` instead of `;` that will ensure the cd succeeds before trying to build the docker image, whereas the semicolon will not.